src/GitHub.InlineReviews/InlineCommentMarginProvider.cs

```diff
@@ -30,68 +30,59 @@ internal sealed class InlineCommentMarginProvider : IWpfTextViewMarginProvider
             IInlineCommentPeekService peekService,
-            IPullRequestSessionManager sessionManager)
+            IPullRequestSessionManager sessionManager,
+            IPackageSettings packageSettings)
         {
             this.editorFormatMapService = editorFormatMapService;
             this.tagAggregatorFactory = tagAggregatorFactory;
             this.peekService = peekService;
             this.sessionManager = sessionManager;
+            this.packageSettings = packageSettings;
         }
 
         public IWpfTextViewMargin CreateMargin(IWpfTextViewHost wpfTextViewHost, IWpfTextViewMargin parent)
         {
             var textView = wpfTextViewHost.TextView;
             var editorFormatMap = editorFormatMapService.GetEditorFormatMap(textView);
             var glyphFactory = new InlineCommentGlyphFactory(peekService, textView, editorFormatMap);
-
-            Func<Grid> gridFactory = () => new GlyphMarginGrid();
-            return CreateMargin(glyphFactory, gridFactory, wpfTextViewHost, parent, editorFormatMap);
+            return CreateMargin(glyphFactory, wpfTextViewHost, parent, editorFormatMap);
         }
 
-        IWpfTextViewMargin CreateMargin<TGlyphTag>(IGlyphFactory<TGlyphTag> glyphFactory, Func<Grid> gridFactory,
+        IWpfTextViewMargin CreateMargin<TGlyphTag>(IGlyphFactory<TGlyphTag> glyphFactory,
             IWpfTextViewHost wpfTextViewHost, IWpfTextViewMargin parent, IEditorFormatMap editorFormatMap) where TGlyphTag : ITag
         {
             var tagAggregator = tagAggregatorFactory.CreateTagAggregator<TGlyphTag>(wpfTextViewHost.TextView);
-            var margin = new GlyphMargin<TGlyphTag>(wpfTextViewHost, glyphFactory, gridFactory, tagAggregator, editorFormatMap,
-                IsMarginVisible, MarginPropertiesName, MarginName, true, 17.0);
 
-            TrackCommentGlyphOnDiffView(wpfTextViewHost, margin.VisualElement);
-            return margin;
-        }
+            var isDiffView = IsDiffView(wpfTextViewHost);
+            var glyphMarginViewModel = new GlyphMarginViewModel(packageSettings, sessionManager, wpfTextViewHost, isDiffView);
+            var glyphMarginView = new GlyphMarginView { DataContext = glyphMarginViewModel };
+            tagAggregator.BatchedTagsChanged += (s, e) => glyphMarginViewModel.RefreshVisibility();
+            glyphMarginView.Unloaded += (s, e) => glyphMarginViewModel.Dispose();
 
-        void TrackCommentGlyphOnDiffView(IWpfTextViewHost host, UIElement marginElement)
-        {
-            var textView = host.TextView;
-            if (textView.Roles.Contains("DIFF"))
-            {
-                var router = new MouseEnterAndLeaveEventRouter<AddInlineCommentGlyph>();
-                router.Add(host.HostControl, marginElement);
-            }
-        }
+            var margin = new GlyphMargin<TGlyphTag>(wpfTextViewHost, glyphFactory, glyphMarginView, tagAggregator, editorFormatMap,
+                MarginPropertiesName, MarginName, true, 17.0);
```

Comment:
Is this the width of the margin? If so, it should be a const value somewhere else, so we don't have to hunt it down in the code.